### PR TITLE
Starlite parameter types

### PR DESCRIPTION
**Pull request by peterschutt**
Refactoring of the application/router/controller/handler parameters for 2.0.

- `route_handlers` optional, defaulting to empty list.
- all `SingleOrList[T]` params in v1 now `Sequence[T]`.
- prefer abstract container types for parameters.

# PR Checklist

- [ ] Have you followed the guidelines in `CONTRIBUTING.rst`?
- [ ] Have you got 100% test coverage on new code?
- [ ] Have you updated the prose documentation?
- [ ] Have you updated the reference documentation?

### CODE REVIEW

File: starlite/controller.py [hunk 1]
```diff
@@ -86,11 +86,11 @@ class Controller:
...
     """A dictionary that maps handler functions to status codes and/or exception types."""
-    guards: Optional[List["Guard"]]
+    guards: Optional[Sequence["Guard"]]
     """A list of :class:`Guard <starlite.types.Guard>` callables."""
-    middleware: Optional[List["Middleware"]]
+    middleware: Optional[Sequence["Middleware"]]
     """A list of :class:`Middleware <starlite.types.Middleware>`."""
-    opt: Optional[Dict[str, Any]]
+    opt: Optional[Mapping[str, Any]]
     """A string key dictionary of arbitrary values that can be accessed in :class:`Guards <starlite.types.Guard>` or wherever
```
Comment (provinzkraut) [inline starlite/controller.py:94]:
```suggestion
    """A string key mapping of arbitrary values that can be accessed in :class:`Guards <starlite.types.Guard>` or wherever
```

File: starlite/controller.py [hunk 2]
```diff
@@ -86,11 +86,11 @@ class Controller:
     """
     exception_handlers: Optional["ExceptionHandlersMap"]
     """A dictionary that maps handler functions to status codes and/or exception types."""
```
Comment (provinzkraut) [inline starlite/controller.py:88]:
```suggestion
    """A map of handler functions to status codes and/or exception types."""
```

File: starlite/controller.py [hunk 3]
```diff
@@ -86,11 +86,11 @@ class Controller:
...
-    guards: Optional[List["Guard"]]
+    guards: Optional[Sequence["Guard"]]
     """A list of :class:`Guard <starlite.types.Guard>` callables."""
```
Comment (provinzkraut) [inline starlite/controller.py:90]:
```suggestion
    """A sequence of :class:`Guard <starlite.types.Guard>` callables."""
```

File: starlite/controller.py [hunk 4]
```diff
@@ -86,11 +86,11 @@ class Controller:
...
-    middleware: Optional[List["Middleware"]]
+    middleware: Optional[Sequence["Middleware"]]
     """A list of :class:`Middleware <starlite.types.Middleware>`."""
```
Comment (provinzkraut) [inline starlite/controller.py:92]:
```suggestion
    """A sequence of :class:`Middleware <starlite.types.Middleware>`."""
```

File: starlite/controller.py [hunk 5]
```diff
@@ -114,9 +114,9 @@ class Controller:
     """A list of [Cookie](starlite.datastructures.Cookie] instances."""
     response_headers: Optional["ResponseHeadersMap"]
     """A string keyed dictionary mapping :class:`ResponseHeader <starlite.datastructures.ResponseHeader>` instances."""
-    tags: Optional[List[str]]
+    tags: Optional[Sequence[str]]
     """A list of string tags that will be appended to the schema of all route handlers under the controller."""
```
Comment (provinzkraut) [inline starlite/controller.py:118]:
```suggestion
    """A sequence of string tags that will be appended to the schema of all route handlers under the controller."""
```

File: starlite/controller.py [hunk 6]
```diff
@@ -114,9 +114,9 @@ class Controller:
...
-    security: Optional[List["SecurityRequirement"]]
+    security: Optional[Sequence["SecurityRequirement"]]
     """A list of dictionaries that to the schema of all route handlers under the controller."""
```
Comment (provinzkraut) [inline starlite/controller.py:120]:
```suggestion
    """A sequence of dictionaries that to the schema of all route handlers under the controller."""
```

File: starlite/handlers/asgi.py [hunk 7]
```diff
@@ -23,12 +23,12 @@ class ASGIRouteHandler(BaseRouteHandler["ASGIRouteHandler"]):
     @validate_arguments(config={"arbitrary_types_allowed": True})
     def __init__(
         self,
-        path: Union[Optional[str], Optional[List[str]]] = None,
+        path: Optional[Union[str, Sequence[str]]] = None,
```
Comment (provinzkraut) [inline starlite/handlers/asgi.py:26]: The types changed here also need to be changed in the docstring, but I can't comment on that

File: starlite/handlers/base.py [hunk 8]
```diff
@@ -72,14 +74,14 @@ class BaseRouteHandler(Generic[T]):
     @validate_arguments(config={"arbitrary_types_allowed": True})
     def __init__(
         self,
-        path: Union[Optional[str], Optional[List[str]]] = None,
+        path: Optional[Union[str, Sequence[str]]] = None,
```
Comment (provinzkraut) [inline starlite/handlers/base.py:77]: The types changed here also need to be changed in the docstring, but I can't comment on that

File: starlite/handlers/http.py [hunk 9]
```diff
@@ -312,7 +319,7 @@ class HTTPRouteHandler(BaseRouteHandler["HTTPRouteHandler"]):
     @validate_arguments(config={"arbitrary_types_allowed": True})
     def __init__(
         self,
-        path: Union[Optional[str], Optional[List[str]]] = None,
+        path: Optional[Union[str, Sequence[str]]] = None,
```
Comment (provinzkraut) [inline starlite/handlers/http.py:322]: The types changed here also need to be changed in the docstring, but I can't comment on that

File: starlite/router.py [hunk 10]
```diff
@@ -94,19 +96,19 @@ def __init__(
         after_response: Optional[AfterResponseHookHandler] = None,
         before_request: Optional[BeforeRequestHookHandler] = None,
         cache_control: Optional[CacheControlHeader] = None,
-        dependencies: Optional[Dict[str, Provide]] = None,
+        dependencies: Optional[Mapping[str, Provide]] = None,
```
Comment (provinzkraut) [inline starlite/router.py:99]: Docstrings need to be adjusted to the new types

File: starlite/testing/create_test_client.py [hunk 11]
```diff
@@ -46,34 +45,34 @@
         OnAppInitHandler,
         ParametersMap,
         ResponseType,
-        SingleOrList,
     )
+    from starlite.types.composite_types import InitialStateType
+    from starlite.types.helper_types import OptionalSequence
 
 
 def create_test_client(
```
Comment (provinzkraut) [inline starlite/testing/create_test_client.py:53]: Docstrings need to be adjusted to the new types

File: starlite/utils/exception.py [hunk 12]
```diff
@@ -34,8 +34,8 @@ def get_exception_handler(exception_handlers: "ExceptionHandlersMap", exc: Excep
     if not exception_handlers:
         return None
     status_code: Optional[int] = getattr(exc, "status_code", None)
-    if status_code in exception_handlers:
-        return exception_handlers[status_code]  # pyright: ignore
+    if status_code is not None and status_code in exception_handlers:
+        return exception_handlers[status_code]
```
Comment (provinzkraut) [inline starlite/utils/exception.py:38]:
```suggestion
    if status_code and (exception_handler := exception_handlers.get(status_code)):
        return exception_handler
``` 
?

**Review by provinzkraut — CHANGES_REQUESTED** [review]
Very nice changes overall!

- Docstrings need to be adjusted to the new types in various places
- The semantic handler decorators in `handlers/http.py` should have the same changes applied as the base handlers

**Review by Goldziher — APPROVED** [review]

**Pull request merged**